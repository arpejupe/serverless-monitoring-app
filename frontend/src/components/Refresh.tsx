import { Auth } from 'aws-amplify';
import React, { useState } from 'react';
import { Button, Container } from 'semantic-ui-react';

const Refresh = () => {
  const api = process.env.REACT_APP_REST_API as string;

  const [loading, setLoading] = useState(false);

  async function onClick() {
    try {
      setLoading(true);
      const credentials = await Auth.currentSession();
      const idToken = credentials.getIdToken().getJwtToken();
      await fetch(`${api}/fanout`, {
        headers: { Authorization: idToken },
        method: 'POST',
      });
    } finally {
      setLoading(false);
    }
  }

  return (
    <Container textAlign="center">
      <Button
        loading={loading}
        onClick={onClick}
        positive={true}
        data-testid="refresh"
      >
        Refresh
      </Button>
    </Container>
  );
};

export default Refresh;
