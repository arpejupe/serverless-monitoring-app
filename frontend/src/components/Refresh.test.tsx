import React from 'react';
import renderer from 'react-test-renderer';
import { render } from 'react-testing-library';
import Refresh from './Refresh';

jest.mock('aws-amplify');

const fetch = jest.fn();
// @ts-ignore
global.fetch = fetch;

const api = 'api';
process.env.REACT_APP_REST_API = api;

jest.useFakeTimers();

describe('Refresh', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders without crashing', () => {
    const { getByText } = render(<Refresh />);
    expect(getByText('Refresh')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<Refresh />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('calls api on button click', async () => {
    const { Auth } = require('aws-amplify');

    const idToken = 'idToken';
    const getJwtToken = jest.fn(() => idToken);
    const getIdToken = jest.fn(() => ({ getJwtToken }));
    const credentials = { getIdToken };
    Auth.currentSession.mockReturnValue(Promise.resolve(credentials));

    const promise = Promise.resolve();
    fetch.mockReturnValue(promise);

    const { getByTestId } = render(<Refresh />);
    const button = getByTestId('refresh');
    button.click();

    await promise;

    expect(Auth.currentSession).toHaveBeenCalledTimes(1);
    expect(getIdToken).toHaveBeenCalledTimes(1);
    expect(getJwtToken).toHaveBeenCalledTimes(1);

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(`${api}/fanout`, {
      headers: { Authorization: idToken },
      method: 'POST',
    });
  });
});
