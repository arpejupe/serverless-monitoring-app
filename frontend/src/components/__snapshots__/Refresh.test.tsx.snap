// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refresh matches snapshot 1`] = `
<div
  className="ui center aligned container"
>
  <button
    className="ui positive button"
    data-testid="refresh"
    onClick={[Function]}
  >
    Refresh
  </button>
</div>
`;
